pandas/_libs/index.pyx

```diff
@@ -585,17 +571,24 @@ cpdef convert_scalar(ndarray arr, object value):
         raise ValueError("cannot set a Timedelta with a non-timedelta "
                          f"{type(value).__name__}")
 
-    if (issubclass(arr.dtype.type, (np.integer, np.floating, np.complex)) and
-            not issubclass(arr.dtype.type, np.bool_)):
+    else:
+        validate_numeric_casting(arr.dtype, value)
+
+    return value
+
+
+cpdef validate_numeric_casting(dtype, object value):
+    # Note: we can't type dtype as cnp.dtype because that cases dtype.type
```

Comment:
Does typing it as np.dtype work? Yea not sure how a cpdef function would choose between them

Reply:
giving it a cython-style type doesnt work (LMK if the comment here is unclear), and if you're suggesting a python-style annotation, my experience has been that mixing and matching those with cython-style is fragile.  (also this could be an EA dtype)

Reply:
Ah sorry I think I got mixed up by the comment. I think saying “define” instead of “type” would be clearer but not worth changing on its own